src/pocketmine/tile/Hopper.php

```diff
@@ -0,0 +1,309 @@
+		$this->inventory = new HopperInventory($this);
+
+		if(!isset($this->namedtag->Items) or !($this->namedtag->Items instanceof ListTag)){
+			$this->namedtag->Items = new ListTag("Items", []);
+			$this->namedtag->Items->setTagType(NBT::TAG_Compound);
+		}
+
+		for($i = 0; $i < $this->getSize(); ++$i){
+			$this->inventory->setItem($i, $this->getItem($i));
+		}
+		$this->namedtag->TransferCooldown = new IntTag("TransferCooldown", 0);
+
+		$this->scheduleUpdate();
+	}
+
+	public function close(){
+		if($this->closed === false){
+			foreach($this->getInventory()->getViewers() as $player){
+				$player->removeWindow($this->getInventory());
+			}
+
+			foreach($this->getInventory()->getViewers() as $player){
+				$player->removeWindow($this->getInventory());
+			}
+			parent::close();
+		}
+	}
+	
+	public function activate(){
+		$this->isPowered = true;
+	}
+	
+	public function deactivate(){
+		$this->isPowered = false;
+	}
+
+	public function canUpdate(){
+		return $this->namedtag->TransferCooldown->getValue() === 0 and !$this->isPowered;
+	}
+
+	public function resetCooldownTicks(){
+		$this->namedtag->TransferCooldown->setValue(8);
+	}
+
+	public function onUpdate(){
+		if(!($this->getBlock() instanceof HopperBlock)){
+			return false;
+		}
+		//Pickup dropped items
+		//This can happen at any time regardless of cooldown
+		$area = clone $this->getBlock()->getBoundingBox(); //Area above hopper to draw items from
+		$area->maxY = ceil($area->maxY) + 1; //Account for full block above, not just 1 + 5/8
+		foreach($this->getLevel()->getChunkEntities($this->getBlock()->x >> 4, $this->getBlock()->z >> 4) as $entity){
+			if(!($entity instanceof DroppedItem)){
+				continue;
+			}
+			if(!$entity->boundingBox->intersectsWith($area)){
+				continue;
+			}
+
+			$item = $entity->getItem();
+			if(!$item instanceof Item){
+				continue;
+			}
+			if($item->getCount() < 1){
+				$entity->kill();
+				continue;
+			}
+
+			if($this->inventory->canAddItem($item)){
+				$this->inventory->addItem($item);
+				$entity->kill();
+			}
+		}
+
+		if(!$this->canUpdate()){ //Hoppers only update CONTENTS every 8th tick
+			$this->namedtag->TransferCooldown->setValue($this->namedtag->TransferCooldown->getValue() - 1);
+			return true;
+		}
+		
+		//Suck items from above tile inventories
+		$source = $this->getLevel()->getTile($this->getBlock()->getSide(Vector3::SIDE_UP));
+		if($source instanceof Tile and $source instanceof InventoryHolder){
+			$inventory = $source->getInventory();
+			$item = clone $inventory->getItem($inventory->firstOccupied());
+			$item->setCount(1);
+			if($this->inventory->canAddItem($item)){
+				$this->inventory->addItem($item);
+				$inventory->removeItem($item);
+				$this->resetCooldownTicks();
+				if($source instanceof Hopper){
+					$source->resetCooldownTicks();
+				}
+			}
+		}
+		
+		//Feed item into target inventory
+		//Do not do this if there's a hopper underneath this hopper, to follow vanilla behaviour
```

Comment:
Yes, it does. That's already implemented. It just skips feeding in this circumstance.